File: scripts/youngcha/generation/electricmotor.js
const rotL = require("libraries/rotpowerlib");


const electricMotor = rotL.torqueExtend(Block, Building, "electric-motor", rotL.baseTypes.torqueGenerator,{
	
	load(){
		this.super$load();
		this.topsprite = Core.atlas.find(this.name + "-top");
		this.overlaysprite = [
			Core.atlas.find(this.name + "-overlay1"),
			Core.atlas.find(this.name + "-overlay2")];
		this.moving = Core.atlas.find(this.name + "-moving");
		this.bottom = Core.atlas.find(this.name + "-bottom");
		this.base = [
			Core.atlas.find(this.name + "-base1"),
			Core.atlas.find(this.name + "-base2")
		];
		this.coil = [
			Core.atlas.find(this.name + "-coil1"),
			Core.atlas.find(this.name + "-coil2")
		];
		this.mbase = Core.atlas.find(this.name + "-mbase");
		
		
	},
	
},{
	
	updatePre()
	{
		this.getGraphConnector("torque graph").setInertia(25);
		this.getGraphConnector("torque graph").setFriction(0.1);
		this.getGraphConnector("torque graph").setMotorForceMult(this.power.graph.getSatisfaction());
	},
	
	draw() {
		let torquegraph = this.getGraphConnector("torque graph");
		let fixedrot = ((this.rotdeg() + 90) % 180) - 90;
		let variant = ((this.rotation + 1) % 4>=2) ? 1 : 0;
		let rotVar = (this.rotation == 1 || this.rotation == 3) ? 1 : 0;
		let shaftrotog = torquegraph.getRotation();
		let shaftRot = variant==1?360-shaftrotog:shaftrotog;
		Draw.rect(electricMotor.bottom, this.x, this.y, 0);
		Draw.rect(electricMotor.base[rotVar], this.x, this.y, 0);
		Draw.rect(electricMotor.coil[rotVar], this.x, this.y, 0);
		Draw.rect(electricMotor.mbase, this.x, this.y, this.rotdeg());
		//speeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeen
		rotL.drawRotRect(electricMotor.moving, this.x, this.y, 24, 14/4, 24, this.rotdeg(), shaftRot, shaftRot+180);
		Draw.rect(electricMotor.overlaysprite[variant], this.x, this.y, this.rotdeg());
		
		Draw.rect(electricMotor.topsprite, this.x, this.y, 0);
        this.drawTeamTop();

	}
	
	
});

electricMotor.rotate = true;
electricMotor.update = true;
electricMotor.solid = true;
electricMotor.consumes.power(4.5);
electricMotor.getGraphConnectorBlock("torque graph").setAccept([0,1,0,
																0,0,0,
																0,1,0,
																0,0,0]);
electricMotor.getGraphConnectorBlock("torque graph").setMaxSpeed(10);
electricMotor.getGraphConnectorBlock("torque graph").setMaxTorque(20);
electricMotor.getGraphConnectorBlock("torque graph").setBaseFriction(0.1);
electricMotor.getGraphConnectorBlock("torque graph").setBaseInertia(25);

File: main/src/unity/world/blocks/power/ElectricMotor.java
package unity.world.blocks.power;

import arc.graphics.g2d.*;
import unity.graphics.*;

import static arc.Core.atlas;

public class ElectricMotor extends TorqueGenerator{
    public final TextureRegion[] overlayRegions = new TextureRegion[2], baseRegions = new TextureRegion[2], coilRegions = new TextureRegion[2];
    public TextureRegion topRegion, movingRegion, bottomRegion, mbaseRegion;

    public ElectricMotor(String name){
        super(name);
        solid = true;
    }

    @Override
    public void load(){
        super.load();
        
        topRegion = atlas.find(name + "-top");
        movingRegion = atlas.find(name + "-moving");
        bottomRegion = atlas.find(name + "-bottom");
        mbaseRegion = atlas.find(name + "-mbase");
        
        for(int i = 0; i < 2; i++){
            baseRegions[i] = atlas.find(name + "-base" + (i + 1));
            overlayRegions[i] = atlas.find(name + "-overlay" + (i + 1));
            coilRegions[i] = atlas.find(name + "-coil" + (i + 1));
        }
    }

    public class ElectricMotorBuild extends TorqueGeneratorBuild{
        @Override
        protected float generateTorque(){
            return power.graph.getSatisfaction();
        }

        @Override
        public void draw(){
            int variant = (rotation + 1) % 4 >= 2 ? 1 : 0;
            int rotVar = rotation % 2 == 1 ? 1 : 0;
            
            float shaftRot = torque().getRotation();
            
            if(variant == 1) shaftRot = 360f - shaftRot;
            
            Draw.rect(bottomRegion, x, y);
            Draw.rect(baseRegions[rotVar], x, y);
            Draw.rect(coilRegions[rotVar], x, y);
            Draw.rect(mbaseRegion, x, y, rotdeg());

            UnityDrawf.drawRotRect(movingRegion, x, y, 24f, 3.5f, 24f, rotdeg(), shaftRot, shaftRot + 180f);
            Draw.rect(overlayRegions[variant], x, y, rotdeg());
            Draw.rect(topRegion, x, y);
            
            drawTeamTop();
        }
    }
}
